### ISSUE TITLE
Compose+Swarm + networking overlay

### ISSUE THREAD

**Issue by dngferreira** [docker/compose#2525]
Hi,
I am trying to build a elasticsearch cluster using docker swarm (1.0), docker compose(1.51) and the network overlay.  
I'm starting with two nodes: elasticsearch_node_1 and elasticsearch_node_2.  
When connecting node_1 to node_2 works great I can just ping elasticsearch_node_1 using it's container name:  
elasticsearch_node_2$ ping elasticsearch_node_1  
but doing 
elasticsearch_node_2$ ping elasticsearch_node_2 
gives an error, it can't resolve the hostname.

Is this a known issue or am I doing some thing wrong?

**Comment by aanand** [comment]
Are you passing `--x-networking` and `--x-network-driver` to `docker-compose`?

What's the output of:

```
$ docker network inspect elasticsearch
$ docker exec elasticsearch_node_1 cat /etc/hosts
$ docker exec elasticsearch_node_2 cat /etc/hosts
```

**Comment by lonangel** [comment]
$ docker network inspect elastic
```
[
    {
        "Name": "elastic",
        "Id": "be7b4a7fdd42977add3944b626df31f8bbdd7e6f73e6920330d8f87b6d67cd3b",
        "Scope": "global",
        "Driver": "overlay",
        "IPAM": {
            "Driver": "default",
            "Config": [
                {}
            ]
        },
        "Containers": {
            "03d31f86bd504e05d459a9bb07df1004007bca6144e74c6e0ddcbda158fa4312": {
                "EndpointID": "c454fecac7ff9550c0dfe8b59d9ce3e3a198c4d7178e2d5de828ad4016f4a685",
                "MacAddress": "02:42:0a:00:00:02",
                "IPv4Address": "10.0.0.2/24",
                "IPv6Address": ""
            }
        },
        "Options": {}
    }
]
```

$ docker exec elastic_2 cat /etc/hosts
```

10.0.0.3        30fcc52f15a4
127.0.0.1       localhost
::1     localhost ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
10.0.0.2        elastic_1
10.0.0.2        elastic_1.elastic
```

$ docker exec elastic_1 cat /etc/hosts
```
10.0.0.2        03d31f86bd50
127.0.0.1       localhost
::1     localhost ip6-localhost ip6-loopback
fe00::0 ip6-localnet
ff00::0 ip6-mcastprefix
ff02::1 ip6-allnodes
ff02::2 ip6-allrouters
10.0.0.3        elastic_2
10.0.0.3        elastic_2.elastic
```

**Comment by dnephin** [comment]
I think that's expected, because the hostname is not actually `elastic_1`, the hostname is `03d31f86bd50`. The other names are just aliases for discovery.

**Comment by lonangel** [comment]
Eventhough they are alias shouldn't the container know its own alias?

**Comment by lonangel** [comment]
Adding the hostname parameter to the yaml file solves my issue but it looks like a hack to do it like this.
```
elastic_1:
  image: elasticsearch
  container_name: elastic_1
  hostname: elastic_1
```
If I understand it correctly this can't be changed on the compose side.
The change would have to be made on the docker swarm source code, right?

**Comment by dnephin** [comment]
Yes, it's handled by libnetwork. I think there are some changes coming in the next version that will make this work.

**Comment by brunoborges** [comment]
+1 for this. The service should know which hostname was assigned to it that is used by other services.

**Comment by jippi** [comment]
:+1:

**Comment by nazar-pc** [comment]
Agree, there should be a way to get own hostname, but placing it into `/etc/hosts` would cause BC break. Maybe some env var can be added for this purpose? In images that create clusters with just scaling it is not trivial to find out how to connect containers by hostnames (IPs are good, but even less reliable than hostnames).

**Comment by dnephin** [comment]
This is fixed with Compose 1.6 (using the v2 format), and Docker 1.10.  With the DNS resolver replacing `/etc/hosts` conatiners are now able to resolve their own hostname and alises.

**Comment by jippi** [comment]
YAY! Good job
